File: apps/portal-company/src/types/coupon.ts
// src/types/coupon.ts

// enum igual ao backend
export const DiscountType = {
  percent: "percent",
  fixed: "fixed",
} as const;

export type DiscountType = keyof typeof DiscountType; // "percent" | "fixed"

// Create: obrigatórios name e code; demais opcionais
export interface CouponCreate {
  name: string;
  code: string;

  description?: string;

  is_active?: boolean;
  is_visible?: boolean;

  usage_limit_total?: number | null;
  usage_limit_per_user?: number | null;

  min_order_amount?: number | null;

  discount_type?: DiscountType | null;
  discount_value?: number | null;

  category_ids?: string[] | null; // UUID strings
  item_ids?: string[] | null;

  source_location_name?: string | null;
  source_lat?: number | null;
  source_lng?: number | null;
}

// Update parcial (todos opcionais)
export interface CouponUpdate {
  name?: string;
  code?: string;

  description?: string | null;

  is_active?: boolean;
  is_visible?: boolean;

  usage_limit_total?: number | null;
  usage_limit_per_user?: number | null;

  min_order_amount?: number | null;

  discount_type?: DiscountType | null;
  discount_value?: number | null;

  category_ids?: string[] | null;
  item_ids?: string[] | null;

  source_location_name?: string | null;
  source_lat?: number | null;
  source_lng?: number | null;
}

export interface CouponRead {
  id: string;         // UUID
  company_id: string; // UUID

  name: string;
  code: string;
  description: string | null;

  is_active: boolean;
  is_visible: boolean;

  usage_limit_total: number | null;
  usage_limit_per_user: number | null;

  min_order_amount: number | null;

  discount_type: DiscountType | null;
  discount_value: number | null;

  category_ids: string[]; // sempre lista
  item_ids: string[];

  source_location_name: string | null;
  source_lat: number | null;
  source_lng: number | null;

  created_at: string | null; // ISO date
  updated_at: string | null; // ISO date
}

export interface PaginatedCoupons {
  total: number;
  skip: number;
  limit: number;
  items: CouponRead[];
}

// ==== Redeem / Validate ====
export interface CouponValidateRequest {
  code: string;
  user_id: string;
  amount: number;
  item_ids?: string[];              // UUIDs dos itens do carrinho
  source_lat?: number;
  source_lng?: number;
  source_location_name?: string;
  dry_run?: boolean;                // preview=true (default no preview)
}

export interface CouponValidateResponse {
  valid: boolean;
  reason?: string | null;
  coupon_id?: string | null;
  discount: number;                 // valor do desconto calculado
  final_amount?: number | null;     // amount - discount (só vem quando valid=true)
  redemption_id?: string | null;    // só quando efetivamente resgata (dry_run=false)
}

File: apps/backend/app/schemas/coupon.py
from datetime import datetime
from enum import Enum
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DiscountType(str, Enum):
    percent = "percent"
    fixed   = "fixed"


# Para criação: só name e code são obrigatórios
class CouponCreate(BaseModel):
    name: str = Field(..., max_length=120)
    code: str = Field(..., max_length=64)

    description: Optional[str] = None

    is_active: Optional[bool] = True
    is_visible: Optional[bool] = True

    usage_limit_total: Optional[int] = None
    usage_limit_per_user: Optional[int] = None

    min_order_amount: Optional[float] = None

    discount_type: Optional[DiscountType] = None
    discount_value: Optional[float] = None

    category_ids: Optional[List[UUID]] = None
    item_ids: Optional[List[UUID]] = None

    source_location_name: Optional[str] = None
    source_lat: Optional[float] = None
    source_lng: Optional[float] = None


# Para update parcial (PATCH-like via PUT, dependendo do seu padrão)
class CouponUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=120)
    code: Optional[str] = Field(None, max_length=64)

    description: Optional[str] = None

    is_active: Optional[bool] = None
    is_visible: Optional[bool] = None

    usage_limit_total: Optional[int] = None
    usage_limit_per_user: Optional[int] = None

    min_order_amount: Optional[float] = None

    discount_type: Optional[DiscountType] = None
    discount_value: Optional[float] = None

    category_ids: Optional[List[UUID]] = None  # se vier, substitui o set
    item_ids: Optional[List[UUID]] = None      # se vier, substitui o set

    source_location_name: Optional[str] = None
    source_lat: Optional[float] = None
    source_lng: Optional[float] = None


class CouponRead(BaseModel):
    id: UUID
    company_id: UUID

    name: str
    code: str
    description: Optional[str]

    is_active: bool
    is_visible: bool

    usage_limit_total: Optional[int]
    usage_limit_per_user: Optional[int]

    min_order_amount: Optional[float]

    discount_type: Optional[DiscountType]
    discount_value: Optional[float]

    category_ids: List[UUID] = []  # retornaremos como lista de IDs
    item_ids: List[UUID] = []

    source_location_name: Optional[str]
    # Para leitura, manteremos como opcionais (podem vir None se não consultarmos ST_X/ST_Y)
    source_lat: Optional[float] = None
    source_lng: Optional[float] = None

    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class PaginatedCoupons(BaseModel):
    total: int
    skip: int
    limit: int
    items: List[CouponRead]

    model_config = ConfigDict(from_attributes=True)
